pgrx/src/memcx.rs

```diff
@@ -45,3 +55,99 @@ where
 
     f(&memcx)
 }
+
+/// Does alignment / padding logic for pre-Postgres16 8-byte alignment.
+#[cfg(all(
+    feature = "nightly",
+    any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")
+))]
+fn layout_for_pre16(
+    layout: std::alloc::Layout,
+) -> Result<std::alloc::Layout, std::alloc::AllocError> {
+    Ok(layout.align_to(8).map_err(|_e| std::alloc::AllocError)?.pad_to_align())
+}
+
+#[cfg(feature = "nightly")]
+unsafe impl<'mcx> std::alloc::Allocator for &MemCx<'mcx> {
+    // Future-proofing - currently only pg16 supports arbitrary alignment, but that is likely
+    // to change, whereas old versions are unlikely to lose 8-byte alignment.
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    fn allocate(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // On versions before Postgres 16, alignment is always 8 byte / 64 bit.
+        let layout = layout_for_pre16(layout)?;
+        unsafe {
+            let ptr = pgrx_pg_sys::MemoryContextAlloc(self.ptr.as_ptr(), layout.size());
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    #[cfg(not(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")))]
+    fn allocate(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
```

Comment:
One thing I don't quite understand.  The returned byte array isn't associated with any lifetime.  Not the lifetime of `&self` nor even `'mcx`.

The doc comments on `std::alloc::Allocator::allocate()` don't make a lot of sense to me, so how does the compiler otherwise ensure that the function's result can't be used after free?

Reply:
For `Vec<T, A>` the lifetime of `A` is what we're using to make it so that allocations cannot outlive their memory context.

Reply:
Yeah it does state 
```
    /// The returned block of memory remains valid as long as it is [*currently allocated*] and the shorter of:
    ///   - the borrow-checker lifetime of the allocator type itself.
    ///   - as long as at the allocator and all its clones has not been dropped.
```

but the problem is that these methods do not take or return lifetimes, so it might just be compiler magic we have to trust? @workingjubilee would know more here, I think.

Reply:
> For `Vec<T, A>` the lifetime of `A` is what we're using to make it so that allocations cannot outlive their memory context.

This sounds right.